File: AFMobile/AFAndroid/app/src/main/java/cz/cvut/fel/matyapav/afandroid/parsers/JSONDefinitionParser.java
package cz.cvut.fel.matyapav.afandroid.parsers;

import android.content.Context;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import cz.cvut.fel.matyapav.afandroid.components.parts.ClassDefinition;
import cz.cvut.fel.matyapav.afandroid.components.parts.FieldInfo;
import cz.cvut.fel.matyapav.afandroid.components.parts.FieldOption;
import cz.cvut.fel.matyapav.afandroid.components.parts.LayoutProperties;
import cz.cvut.fel.matyapav.afandroid.components.parts.ValidationRule;
import cz.cvut.fel.matyapav.afandroid.enums.LabelPosition;
import cz.cvut.fel.matyapav.afandroid.enums.LayoutDefinitions;
import cz.cvut.fel.matyapav.afandroid.enums.LayoutOrientation;
import cz.cvut.fel.matyapav.afandroid.enums.SupportedWidgets;
import cz.cvut.fel.matyapav.afandroid.utils.Constants;

/**
 * @author Pavel Matyáš (matyapav@fel.cvut.cz).
 * @since 1.0.0..
 */
public class JSONDefinitionParser implements JSONParser {

    @Override
    public ClassDefinition parse(JSONObject classInfo) {
        ClassDefinition definition = null;

        try {
            //Parse class name and create data pack with this class name
            System.err.println("PARSING CLASS " + classInfo.getString(Constants.CLASS_NAME));
            definition = new ClassDefinition(classInfo.getString(Constants.CLASS_NAME));

            //Parse layout
            JSONObject layout = classInfo.optJSONObject(Constants.LAYOUT);
            definition.setLayout(createLayoutProperties(layout));

            //Parse fields
            JSONArray fields = classInfo.optJSONArray(Constants.FIELD_INFO);
            if (fields != null) {
                for (int i = 0; i < fields.length(); i++) {
                    JSONObject field = fields.getJSONObject(i);
                    definition.addFieldInfo(parseFieldInfo(field));
                }
            }
            JSONArray innerClasses = classInfo.optJSONArray(Constants.INNER_CLASSES);
            if (innerClasses != null) {
                for (int i = 0; i < innerClasses.length(); i++) {
                    JSONObject innerClass = innerClasses.getJSONObject(i);
                    definition.addInnerClass(parse(innerClass)); //recursion;
                }
            }
        } catch (JSONException e) {
            e.printStackTrace();
        }

        return definition;
    }

    private FieldInfo parseFieldInfo(JSONObject field) throws JSONException {
        System.out.println("PARSING FIELD " + field.getString(Constants.ID));
        FieldInfo fieldInfo = new FieldInfo();
        try {
            String widgetType = field.isNull(Constants.WIDGET_TYPE) ? null : field.getString(Constants.WIDGET_TYPE);
            if (widgetType != null) {
                fieldInfo.setWidgetType(SupportedWidgets.valueOf(widgetType));
            }
        } catch (IllegalArgumentException e) {
            System.err.println(e.getLocalizedMessage());
        }
        fieldInfo.setId(field.getString(Constants.ID));

        fieldInfo.setLabelText(field.isNull(Constants.LABEL) ? null : field.getString(Constants.LABEL));
        fieldInfo.setIsClass(Boolean.valueOf(field.getString(Constants.CLASS_TYPE)));
        fieldInfo.setVisible(Boolean.valueOf(field.getString(Constants.VISIBLE)));
        fieldInfo.setReadOnly(Boolean.valueOf(field.getString(Constants.READ_ONLY)));
        //field layout
        fieldInfo.setLayout(createLayoutProperties(field.getJSONObject(Constants.LAYOUT)));
        //rules
        JSONArray rules = field.optJSONArray(Constants.RULES);
        if (rules != null) {
            for (int i = 0; i < rules.length(); i++) {
                fieldInfo.addRule(createRule(rules.getJSONObject(i)));
            }
        }
        //options
        JSONArray options = field.optJSONArray(Constants.OPTIONS);
        if (options != null) {
            for (int i = 0; i < options.length(); i++) {
                fieldInfo.addOption(createOption(options.getJSONObject(i)));
            }
        }

        return fieldInfo;
    }

    private LayoutProperties createLayoutProperties(JSONObject layoutJson) throws JSONException {
        LayoutProperties layoutProp = new LayoutProperties();
        if (layoutJson == null) {
            layoutProp.setLayoutDefinition(LayoutDefinitions.ONECOLUMNLAYOUT);
            layoutProp.setLayoutOrientation(LayoutOrientation.AXISX);
            layoutProp.setLabelPosition(LabelPosition.BEFORE);
            return layoutProp; //with default values
        }

        String layDefName = layoutJson.isNull(Constants.LAYOUT_DEF) ? null : layoutJson.optString(Constants.LAYOUT_DEF, null);
        if (layDefName != null) {
            LayoutDefinitions layDef = LayoutDefinitions.valueOf(layDefName);
            layoutProp.setLayoutDefinition(layDef);
        }
        String orientation = layoutJson.isNull(Constants.LAYOUT_ORIENT) ? null : layoutJson.optString(Constants.LAYOUT_ORIENT, null);
        if (orientation != null) {
            LayoutOrientation layOrient = LayoutOrientation.valueOf(orientation);
            layoutProp.setLayoutOrientation(layOrient);
        }
        String position = layoutJson.isNull(Constants.LABEL_POS) ? null : layoutJson.optString(Constants.LABEL_POS, null);
        if (position != null) {
            LabelPosition labelPos = LabelPosition.valueOf(position);
            layoutProp.setLabelPosition(labelPos);
        }

        return layoutProp;
    }

    private ValidationRule createRule(JSONObject ruleJson) throws JSONException {
        ValidationRule rule = new ValidationRule();
        rule.setValidationType(ruleJson.getString(Constants.VALIDATION_TYPE));
        rule.setValue(ruleJson.getString(Constants.VALUE));
        return rule;
    }

    private FieldOption createOption(JSONObject optionJson) throws JSONException {
        FieldOption option = new FieldOption();
        option.setKey(optionJson.getString(Constants.KEY));
        option.setValue(optionJson.getString(Constants.VALUE));
        return option;
    }
}

File: AFMobile/AFWindowsPhone/AFWindowsPhone/parsers/JSONDefinitionParser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using AFWindowsPhone.builders.components.parts;
using System.Diagnostics;
using System.Linq.Expressions;
using AFWindowsPhone.enums;
using AFWindowsPhone.utils;
using Windows.Data.Json;

namespace AFWindowsPhone.parsers
{
    class JSONDefinitionParser : JSONParser
    {
        public ClassDefinition parse(JsonObject classInfo)
        {
            ClassDefinition definition = null;

            try
            {
                //Parse class name and create data pack with this class name
                Debug.WriteLine("PARSING CLASS " + Utils.TryToGetValueFromJson(classInfo[Constants.CLASS_NAME]));
                definition = new ClassDefinition((String) Utils.TryToGetValueFromJson(classInfo[Constants.CLASS_NAME]));

                //Parse layout
                JsonObject layout = (JsonObject) Utils.TryToGetValueFromJson(classInfo[Constants.LAYOUT]);
                definition.setLayout(createLayoutProperties(layout));

                //Parse fields
                JsonArray fields = (JsonArray) Utils.TryToGetValueFromJson(classInfo[Constants.FIELD_INFO]);
                if (fields != null)
                {
                    for (int i = 0; i < fields.Count; i++)
                    {
                        JsonObject field = fields[i].GetObject();
                        definition.addFieldInfo(parseFieldInfo(field));
                    }
                }
                JsonArray innerClasses = (JsonArray) Utils.TryToGetValueFromJson(classInfo[Constants.INNER_CLASSES]);
                if (innerClasses != null)
                {
                    for (int i = 0; i < innerClasses.Count; i++)
                    {
                        JsonObject innerClass = (JsonObject) Utils.TryToGetValueFromJson(innerClasses[i]);
                        definition.addInnerClass(parse(innerClass)); //recursion;
                    }
                }
            }
            catch (Exception e)
            {
                Debug.WriteLine(e.StackTrace);
            }

            return definition;
        }

        private FieldInfo parseFieldInfo(JsonObject field)
        {
            Debug.WriteLine("PARSING FIELD " + Utils.TryToGetValueFromJson(field[Constants.ID]));
            FieldInfo fieldInfo = new FieldInfo();
            try
            {
                fieldInfo.setWidgetType(Utils.ValueOf<SupportedWidgets>(typeof(SupportedWidgets), (String) Utils.TryToGetValueFromJson(field[Constants.WIDGET_TYPE])));
            }
            catch (Exception e)
            {
                Debug.WriteLine(e.Message);
            }
            fieldInfo.setId((String) Utils.TryToGetValueFromJson(field[Constants.ID]));

            fieldInfo.setLabelText((String) Utils.TryToGetValueFromJson(field[Constants.LABEL]));
            fieldInfo.setIsClass((Boolean) Utils.TryToGetValueFromJson(field[Constants.CLASS_TYPE]));
            fieldInfo.setVisible((Boolean) Utils.TryToGetValueFromJson(field[Constants.VISIBLE]));
            fieldInfo.setReadOnly((Boolean) Utils.TryToGetValueFromJson(field[Constants.READ_ONLY]));
            //field layout
            fieldInfo.setLayout(createLayoutProperties((JsonObject) Utils.TryToGetValueFromJson(field[Constants.LAYOUT])));
            //rules)
            JsonArray rules = (JsonArray) Utils.TryToGetValueFromJson(field[Constants.RULES]);
            if (rules != null)
            {
                for (int i = 0; i < rules.Count; i++)
                {
                    fieldInfo.addRule(createRule((JsonObject) Utils.TryToGetValueFromJson(rules[i])));
                }
            }
            //add number rule - it is not among others in json 
            try
            {
                if (fieldInfo.getWidgetType().Equals(SupportedWidgets.NUMBERFIELD))
                {
                    ValidationRule numberRule = new ValidationRule();
                    numberRule.setValidationType("NUMBER");
                    numberRule.setValue("");
                    fieldInfo.addRule(numberRule);
                }
            }
            catch (Exception e)
            {
                Debug.WriteLine(e.Message);
            }
            //options
            JsonArray options = (JsonArray) Utils.TryToGetValueFromJson(field[Constants.OPTIONS]);
            if (options != null)
            {
                for (int i = 0; i < options.Count; i++)
                {
                    fieldInfo.addOption(createOption((JsonObject) Utils.TryToGetValueFromJson(options[i])));
                }
            }

            return fieldInfo;
        }


        private LayoutProperties createLayoutProperties(JsonObject layoutJson)
        {
            LayoutProperties layoutProp = new LayoutProperties();
            if (layoutJson == null)
            {
                layoutProp.setLayoutDefinition(LayoutDefinitions.ONECOLUMNLAYOUT);
                layoutProp.setLayoutOrientation(LayoutOrientation.AXISX);
                layoutProp.setLabelPosition(LabelPosition.BEFORE);
                return layoutProp; //with default values
            }

            try
            {
                String layDefName = (String) Utils.TryToGetValueFromJson(layoutJson[Constants.LAYOUT_DEF]);
                LayoutDefinitions layDef = Utils.ValueOf<LayoutDefinitions>(typeof(LayoutDefinitions),layDefName);
                if (layDef != null)
                {
                    layoutProp.setLayoutDefinition(layDef);
                }

                String orientation = (String) Utils.TryToGetValueFromJson(layoutJson[Constants.LAYOUT_ORIENT]);
                LayoutOrientation layOrient = Utils.ValueOf<LayoutOrientation>(typeof(LayoutOrientation),orientation);
                if (layOrient != null)
                {
                    layoutProp.setLayoutOrientation(layOrient);
                }

                String position = (String) Utils.TryToGetValueFromJson(layoutJson[Constants.LABEL_POS]);

                LabelPosition labelPos = Utils.ValueOf<LabelPosition>(typeof(LabelPosition),position);
                if (labelPos != null)
                {
                    layoutProp.setLabelPosition(labelPos);
                }
            }
            catch (Exception e)
            {
                Debug.WriteLine(e.Message);
                //e.printStackTrace();
            }
            return layoutProp;
        }

        private ValidationRule createRule(JsonObject ruleJson)
        {
            ValidationRule rule = new ValidationRule();
            rule.setValidationType((String) Utils.TryToGetValueFromJson(ruleJson[Constants.VALIDATION_TYPE]));
            rule.setValue((String) Utils.TryToGetValueFromJson(ruleJson[Constants.VALUE]));
            return rule;
        }

        private FieldOption createOption(JsonObject optionJson)
        {
            FieldOption option = new FieldOption();
            option.setKey((String) Utils.TryToGetValueFromJson(optionJson[Constants.KEY]));
            option.setValue((String) Utils.TryToGetValueFromJson(optionJson[Constants.VALUE]));
            return option;
        }

    }
}
